### PR TITLE
fix(security): harden auth & TLS secret, certificate, and admission guards (FB-896)

### PR DESCRIPTION
**Pull request by kop**
## Summary

Security hardening for the auth and TLS work merged in #69, taking the findings from @niryi's review in #79 and landing them against `main` with the defects in that branch fixed.

Closes: FB-896

## Changes

Seven commits: @niryi's two, then five of repairs.

**The findings were valid** — I verified each against the merged code rather than the description:

* **Secret guards were scoped per component.** A `spec.gateway.template` or `spec.metadata.template` sidecar could alias the admin password or a JWT signing key, and one engine could reach a sibling's per-generation serving key. Replaced with a single Instance-wide predicate. My original cross-engine test asserted that access was *acceptable* — that reasoning was wrong, since the sidecar reads the Secret regardless of whose it is.
* **A pre-created Secret could supply the signing key.** `rotationPolicy: Never` plus deterministic Secret names means cert-manager reuses a private key already present in the target, so namespace `create secrets` alone was enough to choose the Instance's JWT signing key. Now gated on cert-manager's certificate-name annotation.
* **100-year certificates with no rotation.** Bounded to 90d/30d with `rotationPolicy: Always` on the two TLS listeners. The signing keypair keeps `Never` — its rotation is operator-coordinated — but gains an explicit `digitalSignature` usage so it cannot inherit serverAuth/clientAuth from cert-manager's defaults.
* **The engine TLS anchor requested a namespace wildcard** with serverAuth, for a certificate whose key is mounted nowhere. Narrowed to one non-routable name.
* **Admission gaps:** `secretRef` as an unscreened second route to protected Secrets, `clientCASecretRef` silently ignored on a disabled listener, two OIDC providers mapping to the same identity, and RSA sizes cert-manager rejects passing admission and then wedging an immutable field.

**Three defects fixed in the third commit:**

* **The gateway→engine TLS version range was empty.** `tls_minimum_protocol_version: TLSv1_3` was set without touching the maximum, and Envoy defaults a *client* context's maximum to TLSv1_2 — so BoringSSL refused every upstream connection with `NO_SUPPORTED_VERSIONS_ENABLED` and the gateway answered every query 503. This was reproducible in #79's CI on both variants.
* **The OIDC username-mapping rule never ran.** `validateUsernameMapping` was reachable only from its own tests, so the duplicate-identity check was inert in admission. Now invoked per provider against the resolved admin username.
* **Dead code from the removed key-parameter freeze** — an orphaned helper and a doc comment for a function that no longer exists.

**Three more in the fourth**, from the review agents on this PR — all three valid:

* **The signing-key provenance check did not establish provenance.** `cert-manager.io/certificate-name` is ordinary metadata: whoever creates the Secret sets it, and cert-manager rewrites the same value after issuing, so a planted key satisfies the check. Replaced with creation-order comparison, the one signal here the API server takes away from the client. Both halves confirmed against a live apiserver, not reasoned about.
* **The OIDC uniqueness guard tested for the substring `iss`**, which `{{ swiss_id }}` and `{{ dismissal }}` satisfy while referencing nothing issuer-related. Now matches a template reference to the claim.
* **`spec.tls.gateway.secretRef` could name a per-generation engine serving Secret.** Those never appear in the Instance's status, so the exact-name screen missed them; the payoff is an engine private key in a pod whose sidecars the template author controls, which is enough to impersonate an engine to the gateway. Now uses the same shape match the pod-template guards do.

**And one more in the fifth**, also from the review agents and also valid: **signing keys were protected only once `status.auth.signingKeys` named them.** Template validation runs before the key is provisioned, so on a first apply the bootstrap name was unprotected and a template aliasing it was admitted and rendered, yielding real key material the moment cert-manager issued. Rotation generations had the same gap. Now matched by name shape, reserving the whole `<instance>-auth-signing` prefix, across all three gates.

**And one in the seventh:** the signing-key shape match was scoped to the Instance being validated, so a template on one Instance could name a **sibling's** signing key out of the same namespace. Same escalation the guard exists to stop, one level up — the engine serving-certificate match was already any-engine for exactly that reason. Now matched across every Instance.

## Known limitation

The signing-key provenance check narrows the hole it was written for; it does not close it. The creation-order comparison anchors on the Certificate, and `applySSA` is server-side apply with `ForceOwnership`, which adopts a pre-existing object — so a principal who can create `Certificate` objects as well as Secrets creates the Certificate first, plants the Secret after it, and the ordering holds honestly.

What ships here closes the `create secrets`-only path, which was the reported vulnerability. `verifySigningKeyProvenance`'s doc comment and `docs/security.mdx` both state the boundary, and the guidance is to treat `create certificates` in a namespace running Firebolt Instances as equivalent to trusting the holder with the JWT signing key.

Screening also cannot reach a sibling Instance's **admin password** Secret: the user names that one, so no shape recognizes it, and it is screened by exact name against the Instance under review. Two Instances in a namespace using different names for it can reach each other's. Documented; closing it needs the predicate to enumerate every Instance in the namespace, which admission has no client for.

Two fixes were built and discarded rather than shipped half-working: a Certificate-UID witness (breaks rotation, because a pending mint's UID is discarded before it can be persisted — caught by the rotation-lifecycle and TLA+ state-cover tests) and `rotationPolicy: Always` for signing (closes it properly, but regenerates the key under a stable kid at renewal, which is the `NoValidationGap` violation the rotation state machine exists to prevent). Tracked as **FB-2563**; it is a design decision to model in the TLA+ spec, not a patch.

## Verification

* [x] Automated tests
* [x] Manual verification

Details:

```text
go build ./...            clean
make test                 4635 tests, 9 packages, all pass
make lint                 0 issues  (#79 had 14)
make manifests generate   idempotent
make formal-verify        fixtures current  (#79 had a hand-reformatted generated file)
make formal-check         all three specs pass TLC
make formal-check-counterexample
                          OK: naive retain-window anchor still violates NoValidationGap
make -C docs check        passes
go vet -tags e2e          clean

Every guard checked for vacuity by disabling it and confirming a test fails:

  creationTimestamp ordering disabled       -> TestVerifySigningKeyProvenance FAIL
  iss regex reverted to substring           -> TestValidateUsernameMapping FAIL
  engine-gen shape match disabled           -> TestValidateTLS_Forbids...     FAIL
  signing prefix match disabled             -> TestSigningSecretProtected...  FAIL
                                               (all 6 assertions: 3 gates x bootstrap/rotated)

creationTimestamp semantics, probed against a live apiserver via envtest —
this is the assumption the provenance fix rests on:

  submitted creationTimestamp: 2000-01-01 00:00:00 UTC
  read back:                   2026-07-27 09:39:56 UTC   (server-assigned)
  submitted annotation:        "forged"
  read back:                   "forged"                  (client-controlled)

Upstream TLS range, reproduced against Envoy v1.37.2 with the transport
socket this operator renders, against a server capped at 1.2 and one
allowing 1.3:

  min 1.3, no max (as in #79)   engine max 1.2  ->  503 NO_SUPPORTED_VERSIONS_ENABLED
  min 1.3, no max (as in #79)   engine max 1.3  ->  503 NO_SUPPORTED_VERSIONS_ENABLED
  min 1.2 / max 1.3 (this PR)   engine max 1.2  ->  200, negotiated TLS 1.2
  min 1.2 / max 1.3 (this PR)   engine max 1.3  ->  200, negotiated TLS 1.3

The failure reproduces even against a 1.3-capable server, so it was the empty
range and never engine capability. That run also exercises the narrowed SAN
matcher, which #79's CI never reached.
```

The E2E `carries the token through the gateway to an engine` spec is the gate that matters here: it is what caught the original `BAD_ECC_CERT` defect on the same hop, and it is what failed on #79.

## Risk and rollout

**Risk:** Moderate, concentrated in certificate lifetime. TLS certificates drop from 100 years to 90 days and their keys now rotate on reissue, so existing instances reissue on the next reconcile and roll through the normal path — which is fail-closed on a tightening gateway transition. #69 already made a reissue observable on both hops (the gateway folds each mounted TLS Secret's `resourceVersion` into its config hash; an engine generation rolls on serving-cert fingerprint drift), which is what makes bounded lifetimes safe to adopt now.

**Rollout:** No action required. `duration`/`renewBefore` on `CertManagerSpec` and `crlSecretRef` on `TLSListenerSpec` are additive CRD surface. The key algorithm/size CEL freeze is removed along with its webhook twin — it existed only because `Never` meant cert-manager would never regenerate the key to match, and that reason is gone.

**Rollback:** Revert the branch. The provenance check is the one item that changes behaviour for an *existing* Secret: an Instance whose signing Secret lacks cert-manager's certificate-name annotation will refuse to treat it as its own key material. That is the intended failure mode, but it is the thing to watch on first rollout.

## Reviewer notes

* **Start with:** `fix(security): scope secret guards…` for the guard consolidation and the provenance check — that is the substance. The third commit is repair work.
* **Pay special attention to:** two places. The upstream `tls_params` block in `buildDFPUpstreamTLSTransportSocket` — both ends of the version range have to be set together, setting only the floor is what broke #79, and the comment there says so, so nobody re-introduces it. And `verifySigningKeyProvenance`, where the two documented limits (equal-second collisions pass; a recreated Certificate is admitted on a witnessed fingerprint) are the deliberate edges of the check.
* **Intentionally out of scope:** the `.legitignore` `value_` suppression @niryi flagged. It is repo-wide and history-wide and the digest cannot be confirmed against the string the comment names. I added those entries — they suppress two throwaway e2e admin passwords that no longer exist in the tree but still surface from PR history. Narrowing them is worth doing separately, not blind.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/controller/instance_tls.go [hunk 1]
```diff
@@ -265,6 +334,51 @@ func engineTLSSecretReady(secret *corev1.Secret) bool {
...
+// during a gateway issuer rotation (a supported operation: the CRD leaves the
+// gateway issuer mutable precisely so a client cert can be reissued under a new
+// CA), because the Secret carries the old issuer's annotations until the reissue
+// lands.
+//
+// Deliberately NOT auto-remediated by deleting the offending Secret: destroying key
+// material on the strength of an annotation mismatch is the wrong default, and a
+// genuine mismatch always needs a human to work out where the Secret came from.
+// The operator refuses to use it, reports why, and waits.
+func verifyCertManagerIssued(secret *corev1.Secret, certName string) error {
```
[the remark this answers is not in the document]
Comment (kop) [reply internal/controller/instance_tls.go:371]: Valid, and the sharpest finding on this PR. Confirmed against a live apiserver rather than reasoned about — a Secret created with `cert-manager.io/certificate-name` already set reads back with the attacker's value intact, and cert-manager rewrites the same value after it issues, so a planted key looks correct from every angle a client can control. `verifyCertManagerIssued` was a wrong-Secret filter wearing an authenticity check's doc comment.

Fixed in e307331 by comparing creation order instead. The same probe showed the API server overwrites a client-supplied `creationTimestamp` (submitted `2000-01-01`, read back `2026-07-27`), so it is the one field here the client does not control. cert-manager cannot write a target Secret before its Certificate exists, so a signing Secret older than its Certificate was planted.

```
client-supplied creationTimestamp: 2000-01-01 00:00:00 +0000 UTC
server-returned creationTimestamp: 2026-07-27 09:39:56 +0000 UTC
annotation round-tripped as:       "forged"
```

Two limits are documented rather than papered over: `creationTimestamp` is second-granular so equal-second must pass (the honest path collides there), leaving a sub-second race that also requires knowing the kid; and a Certificate deleted and recreated leaves a Secret that legitimately predates it, allowed only when its public key matches the fingerprint already witnessed in status. Rejection raises an event and waits rather than deleting the Secret, which would destroy the key.

Scoped to the signing keypair. The TLS listeners use `rotationPolicy: Always`, so cert-manager generates a fresh key on every issuance and can never adopt planted material; the annotation check is sufficient there.

File: api/v1alpha1/fireboltinstance_webhook.go [hunk 2]
```diff
@@ -673,6 +682,57 @@ func validatePositiveDurationField(path *field.Path, value string) *field.Error
...
+//     packdb validates each provider in isolation and cannot see this.
+//
+//  2. A mapping that is a bare literal equal to the admin username resolves an
+//     external identity onto the Instance admin account, which packdb re-provisions
+//     from config on every engine start. Only a literal is detectable at admission
+//     (a template's output depends on claims the operator never sees), so this
+//     catches the unambiguous case and leaves the templated one to documentation.
+func validateUsernameMapping(mapping string, path *field.Path, providerCount int, adminName string) field.ErrorList {
+	var errs field.ErrorList
+	if providerCount > 1 && !strings.Contains(mapping, issuerClaim) {
```
[the remark this answers is not in the document]
Comment (kop) [reply api/v1alpha1/fireboltinstance_webhook.go]: Valid — `{{ swiss_id }}` and `{{ dismissal }}` both pass a `strings.Contains(mapping, "iss")` test while referencing nothing issuer-related, which admits exactly the cross-provider collision the rule exists to reject.

Fixed in e307331: the check now matches a template reference to the claim (`\{\{\s*iss\s*\}\}`). Your two bypass strings are in the test table verbatim, alongside `{{ email }}@issuer.example.com`, bare `iss`, and `{{ ISS }}`. Reverting to the substring test fails that test, so the fix is not vacuous.

File: api/v1alpha1/operatorauthority.go [hunk 3]
```diff
@@ -865,9 +897,79 @@ func EngineMountedSecretNames(inst *FireboltInstance) []string {
...
+	}
+	set := make(map[string]struct{}, len(names))
+	for _, n := range names {
+		set[n] = struct{}{}
+	}
+	return func(name string) bool {
+		_, hit := set[name]
+		return hit
+	}
+}
```
[the remark this answers is not in the document]
Comment (kop) [reply api/v1alpha1/operatorauthority.go:973]: Valid. `InstanceProvisionedSecretNames` is built from the Instance's own status, and per-generation engine serving Secrets never appear there, so `spec.tls.gateway.secretRef: <engine>-gN-engine-tls` was admitted. That mounts an engine's private key into the gateway pod, whose sidecars the template author controls, and the gateway authenticates engines on the engine CA plus a namespace wildcard SAN — so holding any engine's key is enough to impersonate an engine to it.

Fixed in e307331 by applying the same shape match the pod-template guards already use, as you suggested rather than inventing a second convention. Two details worth flagging:

* The predicate previously returned `nil` (protecting nothing) when the Instance had provisioned nothing yet. That had to go: engine Secrets exist independently of this Instance's status, so the shape match has to hold on a first apply too. There's a test for that case specifically.
* Admission and the controller live in packages that cannot import each other, so the shape match is now defined once in `api/v1alpha1` and the controller delegates to it. `TestIsGeneratedEngineTLSSecretName_MatchesControllerNaming` feeds real `genResourceName(engine, gen, SuffixEngineTLS)` output through the admission predicate, so a rename on either side fails rather than silently unprotecting every engine serving key.

Also worth recording for anyone reading this thread: this route is not webhook-only. The controller re-runs `ValidateTLS` in both TLS steps, so the fix closes it on a default install with the webhook off, not just at admission.

File: api/v1alpha1/fireboltinstance_webhook.go [hunk 4]
```diff
@@ -673,6 +682,57 @@ func validatePositiveDurationField(path *field.Path, value string) *field.Error
...
+		errs = append(errs, field.Invalid(path, mapping,
+			fmt.Sprintf("must incorporate the %q claim when more than one OIDC provider is configured: "+
+				"a mapping that is not provider-unique lets an identity from one provider resolve to "+
+				"another provider's Firebolt user (for example %q)", issuerClaim, "{{ iss }}|{{ sub }}")))
+	}
```
[the remark this answers is not in the document]
Comment (kop) [reply api/v1alpha1/fireboltinstance_webhook.go:735]: Valid — `{{ swiss_id }}` and `{{ dismissal }}` both pass a `strings.Contains(mapping, "iss")` test while referencing nothing issuer-related, which admits exactly the cross-provider collision the rule exists to reject.

Fixed in e307331: the check now matches a template reference to the claim (`\{\{\s*iss\s*\}\}`). Your two bypass strings are in the test table verbatim, alongside `{{ email }}@issuer.example.com`, bare `iss`, and `{{ ISS }}`. Reverting to the substring test fails that test, so the fix is not vacuous.

File: internal/controller/instance_controller.go [hunk 5]
```diff
@@ -549,45 +553,72 @@ func validateTemplateSecretRefs(
...
+		seen[n] = struct{}{}
+		out = append(out, n)
 	}
-	if len(set) == 0 {
-		return nil
-	}
-	return func(name string) bool {
-		_, hit := set[name]
-		return hit
+	for _, n := range computev1alpha1.InstanceOperatorSecretNames(inst) {
```
[the remark this answers is not in the document]
Comment (kop) [reply internal/controller/instance_controller.go:609]: Valid, and it reaches further than the report. Confirmed the ordering directly: `validateInstanceTemplates` runs at `instance_controller.go:200` and `ensureAuth` at `:217`, so on a first apply `status.auth` is nil and the bootstrap name is in no protected set.

Two things to add:

**Rotation generations have the same gap.** Names are `signingCertificateName(instance, kid)` — `<instance>-auth-signing` for the bootstrap key and `<instance>-auth-signing-<kid>` after that. A future kid's name exists nowhere until the moment it is minted, so a name-only screen can never cover it either. Fixing only the bootstrap name would have left the rotating set open.

**All three gates were affected, not just template validation.** The Instance template gate, the engine template gate, and `spec.tls.*.secretRef` all resolve through predicates built from recorded names. The engine one rebuilds its predicate on the far side of `InstanceInfo`, so the Instance name had to be carried across for the shape match to work there.

Fixed in 25eb512 by reserving the whole `<instance>-auth-signing` prefix, which holds from the first apply and covers every kid. Disabling the match fails the new test on all six combinations (three gates x bootstrap/rotated):

```
instance template gate leaves "inst-auth-signing" unprotected before status exists
engine template gate leaves "inst-auth-signing" unprotected before status exists
spec.tls.gateway.secretRef accepted "inst-auth-signing" before status existed
instance template gate leaves "inst-auth-signing-signing-7" unprotected before status exists
engine template gate leaves "inst-auth-signing-signing-7" unprotected before status exists
spec.tls.gateway.secretRef accepted "inst-auth-signing-signing-7" before status existed
```

One deliberate consequence: reserving the prefix also claims names the operator does not currently use, such as `<instance>-auth-signing-mine`. That is the intended trade and it is now documented in `security.mdx` alongside the engine serving-cert shape, so users know not to put their own Secrets there.

File: internal/controller/instance_tls.go [hunk 6]
```diff
@@ -265,6 +334,117 @@ func engineTLSSecretReady(secret *corev1.Secret) bool {
...
+//
+// A rejection is not auto-remediated. Deleting the Secret would destroy the
+// signing key, and if the material is genuine that retires every token in flight.
+func verifySigningKeyProvenance(
+	secret *corev1.Secret, cert *certmanagerv1.Certificate, witnessedFingerprint *string,
+) error {
+	if err := verifyCertManagerIssued(secret, cert.Name); err != nil {
+		return err
+	}
+	if !secret.CreationTimestamp.Before(&cert.CreationTimestamp) {
```
[the remark this answers is not in the document]
Comment (kop) [reply internal/controller/instance_tls.go:446]: Valid. Confirmed the mechanism rather than reasoning about it: `applySigningCertificate` reaches the Certificate through `applySSA`, which is `client.Apply` with `ForceOwnership` — it adopts a pre-existing object instead of failing, and `creationTimestamp` is fixed at original creation. So the Certificate is not a trusted reference point, and a check measured against it inherits whoever created it first. Create the Certificate at t0, plant the Secret at t1 > t0, and the ordering holds honestly.

**I am not shipping another variant of this check in #89.** I tried two fixes and neither survived, so the honest disposition is a narrowed guard with the residual stated, plus a tracked follow-up: **FB-2563**.

**Attempt 1 — record the Certificate UID from the operator's own create call.** Server-assigned, unforgeable, correct in principle. Implemented and reverted: `mintNextSigningKey` deliberately does not append a key to `Status.Auth.SigningKeys` until its Secret is ready, so the UID recorded at create time is discarded before it can be persisted. The next reconcile rebuilds the candidate with no UID, finds the now-issued Secret, and refuses — rotation breaks entirely. My own rotation-lifecycle and TLA+ state-cover tests caught it. Making it work needs a place to persist a pending mint's UID, which touches the rotation state machine, its TLA+ model, and the generated state-cover fixture.

It also would not have closed the hole. An attacker watching for the operator's own `Create` and planting the Secret before cert-manager's keymanager writes wins that ordering honestly too.

**Attempt 2 — `rotationPolicy: Always` for signing.** The only approach that closes this rather than narrowing it, since cert-manager then never adopts pre-existing key material and provenance stops mattering. Rejected because renewal at day 245 (`duration: 365d` / `renewBefore: 120d`) regenerates the key under a stable kid outside the rotation state machine. Engines reconverge via `authHash`, which folds the public-key fingerprint, but mid-roll an engine holding the old key rejects tokens signed with the new one — precisely the `NoValidationGap` violation `formal/SigningKeyRotation.tla` proves the state machine prevents. That is a design trade to model, not a patch to land on this PR.

**What #89 now says instead.** `0cb9da8` corrects the overstatement in both `verifySigningKeyProvenance`'s doc comment and `docs/security.mdx`. The check closes the `create secrets`-only path — that principal cannot make their Secret postdate an object they must wait for the operator to create — and does not close the `create secrets` + `create certificates` path. The docs now say to treat `create certificates` in a namespace running Firebolt Instances as equivalent to trusting the holder with the JWT signing key.

Worth noting the shape of this thread, because you caught it twice: I claimed a check closed a hole it only narrowed, first with the annotation and then with the timestamp. The comment and the docs now state the boundary instead of implying there isn't one.

File: api/v1alpha1/operatorauthority.go [hunk 7]
```diff
@@ -865,9 +897,134 @@ func EngineMountedSecretNames(inst *FireboltInstance) []string {
...
+		set[n] = struct{}{}
+	}
+	return func(name string) bool {
+		if name == "" {
+			return false
+		}
+		if _, hit := set[name]; hit {
+			return true
+		}
+		return IsGeneratedEngineTLSSecretName(name) || IsInstanceSigningSecretName(inst.Name, name)
```
[the remark this answers is not in the document]
Comment (kop) [reply api/v1alpha1/operatorauthority.go]: Valid. Fixed in `03d7e14`.

This is the cross-component escalation @niryi originally found, one level up: two Instances in a namespace, and whoever can edit Instance A's pod templates mounts `B-auth-signing` into an A sidecar they control, with no Secret read access anywhere. Then they mint tokens B's engines honor.

The inconsistency was mine. `IsGeneratedEngineTLSSecretName` is deliberately any-engine, with a doc comment saying why — "a template must not reach a SIBLING engine's serving key either" — and I wrote the signing match scoped to `inst.Name` right next to it. Same guard, same reasoning, and I only applied it on one side.

`IsSigningKeySecretName(name)` now matches by shape across every Instance in the namespace. The mutation check: scoping it back to the Instance under review fails the new test on both the bootstrap and rotated sibling names. Boundary cases are pinned too, so the broader shape does not start claiming unrelated Secrets — `app-auth-signingkey`, bare `auth-signing`, and `-auth-signing` are all still someone else's.

A side benefit: matching by shape removes the need to carry the Instance name across `InstanceInfo` to rebuild the predicate on the engine side, so the field I added in `25eb512` is gone again.

**One residual, stated rather than implied.** Screening cannot reach a sibling Instance's **admin password** Secret. That one the user names, so no shape recognizes it, and it is screened by exact name against the Instance being validated only. If two Instances in a namespace use different names for their admin password Secret, a template on one can still reference the other's. Recorded in `docs/security.mdx` next to the existing statement that template edit rights amount to namespace Secret read rights, with the recommendation to separate unrelated Instances by namespace. Closing it properly would mean the predicate enumerating every FireboltInstance in the namespace, which admission has no client for today.

**Review by gm42 — APPROVED** [review]
